storage/src/mmr/journaled.rs

```diff
@@ -327,4 +529,145 @@ mod tests {
+                partition_prefix: "test_partition_unpruned".into(),
+                items_per_blob: 7,
+            };
+            let mut mmr = Mmr::<_, _, Sha256>::init(context.clone(), cfg2)
+                .await
+                .unwrap();
+            let mut hasher = Sha256::new();
+            let mut leaves = Vec::with_capacity(LEAF_COUNT);
+            let mut positions = Vec::with_capacity(LEAF_COUNT);
+            for i in 0..LEAF_COUNT {
+                let digest = test_digest(i);
+                leaves.push(digest);
+                let last_leaf = leaves.last().unwrap();
+                let pos = mmr.add(&mut hasher, last_leaf);
+                positions.push(pos);
+                pruned_mmr.add(&mut hasher, last_leaf);
+            }
+            assert_eq!(mmr.size().await.unwrap(), 3994);
+            assert_eq!(pruned_mmr.size().await.unwrap(), 3994);
+
+            // Prune the MMR in increments of 10 making sure the journal is still able to compute
+            // roots and accept new elements.
+            let mut hasher = Sha256::new();
+            for i in 0usize..300 {
+                pruned_mmr.prune(i as u64 * 10).await.unwrap();
+                let digest = test_digest(LEAF_COUNT + i);
+                leaves.push(digest);
+                let last_leaf = leaves.last().unwrap();
+                let pos = pruned_mmr.add(&mut hasher, last_leaf);
+                positions.push(pos);
+                mmr.add(&mut hasher, last_leaf);
+                assert_eq!(pruned_mmr.root(&mut hasher), mmr.root(&mut hasher));
+            }
+
+            // Sync the MMRs.
+            pruned_mmr.sync().await.unwrap();
+            assert_eq!(pruned_mmr.root(&mut hasher), mmr.root(&mut hasher));
+
+            // Close the MMR & reopen.
+            pruned_mmr.close().await.unwrap();
+            let mut pruned_mmr = Mmr::<_, _, Sha256>::init(context.clone(), cfg.clone())
+                .await
+                .unwrap();
+            assert_eq!(pruned_mmr.root(&mut hasher), mmr.root(&mut hasher));
+
+            // Prune everything.
+            pruned_mmr.prune_all().await.unwrap();
+            assert_eq!(pruned_mmr.root(&mut hasher), mmr.root(&mut hasher));
+            let oldest_retained = pruned_mmr.oldest_provable_pos().await.unwrap().unwrap();
+
+            // Close MMR without syncing and make sure state is as expected on reopening.
+            pruned_mmr.close().await.unwrap();
+            let pruned_mmr = Mmr::<_, _, Sha256>::init(context.clone(), cfg.clone())
+                .await
+                .unwrap();
+            assert_eq!(pruned_mmr.root(&mut hasher), mmr.root(&mut hasher));
+            assert_eq!(
+                pruned_mmr.journal.oldest_retained_pos().await.unwrap(),
+                Some(oldest_retained)
+            );
+        });
+    }
+
+    /// Simulate partial writes after pruning, making sure we recover to a valid state.
+    #[test_traced("WARN")]
```

Comment:
I looked into this, and it is technically syntactically incorrect to put an outer attribute between the docstring and the function. What you did above with `simulate_partial_sync` is the right way to format docstrings (comments) + functions.

Reply:
thx, fixed.